design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,254 @@
+
+## Proposals
+
+### Output roles
+Each output role would be stored in a configuration file which suffix is a type name.
+
+Files and Forwards output configuration files in /etc/rsyslog.d:
+```
+  10-output-modules.conf
+  30-output.files
+  30-output.forwards
+```
+
+### Ruleset to switch the output
+Using rsyslog rulesets, we prepare all possible output combinations in the ruleset configuration file as follows.  
+Note: The ruleset name is made from the output types in the alphabetical order.
+```
+==> 50-default-rulesets.conf <==
+# Ansible managed
+
+#
+# Rules for the local system logs
+#
+ruleset(name="files") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+}
+ruleset(name="forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+ruleset(name="files_forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+
+```
+In this PR, just files and forwards outputs are supported.  As described in [To Do's](#to-dos), elasticsearch is going to be added next.
+
+### Input roles
+Based upon the inventry file that users provide, input config jumps to the ruleset.  
+
+The next `imfile` example shows log messages read from /var/log/containers/*.log by the `imfile` plugin are logged in the local files as well as forwarded to the remote host.  
+
+**`imfile` example**
+```
+input(type="imfile" file="/var/log/containers/*.log" tag="container" ruleset="files_forwards")
+```
+The next `imjournal` example shows log messages from journald by the `imjournal` plugin are logged in the local files.
+
+**`imjournal` example**
+```
+module(load="imjournal"
+       StateFile="/var/lib/rsyslog/imjournal.state"
+       RateLimit.Burst="300000"
+       RateLimit.Interval="30"
+       PersistStateInterval="3000")
+call files
+```
+Please note that `imfile` input supports `ruleset` parameter, thus the parameter is used to specify which ruleset to call.
+Other plugins which do not have such a parameter, e.g., `imjournal`, use the `call` statement.
+
+### Inventory file changes
+Here is an example that the inputs from `imjournal` (basics) are logged in the local file;
+the inputs from `imfile` (files) are forwarded to the remote host in addition to the local file,
+
+**Proposed style**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+  - name: file-output
+    type: files
+    logs_collections:
+logs_collections:
+  - name: basic-input
+    type: basics
+    rsyslog_use_files_ruleset: true
+  - name: file-input
+    type: files
+    rsyslog_use_files_ruleset: true
+    rsyslog_use_forwards_ruleset: true
+```
+In this PR, I propose to make the logs_collections (input roles) separate from logging_outputs (output roles) and specify the relationship using boolean parameters rsyslog_use_files_ruleset and/or rsyslog_use_forwards_ruleset.
```

Comment:
Is whether or not to use a `ruleset` an implementation detail?  That is, what difference does it make to the user if the input is sent to the output using a `ruleset` or some other mechanism?  That is, I would think the user would just want to specify "collect logs from these inputs, and send them to these outputs" - and that can be done without the `_ruleset` suffix e.g. `rsyslog_use_files` instead of `rsyslog_use_files_ruleset`

Reply:
> Is whether or not to use a `ruleset` an implementation detail? That is, what difference does it make to the user if the input is sent to the output using a `ruleset` or some other mechanism? 

Ahhh, there's no **other** mechanism for the output...  Only via rulesets.

> That is, I would think the user would just want to specify "collect logs from these inputs, and send them to these outputs" - and that can be done without the `_ruleset` suffix e.g. `rsyslog_use_files` instead of `rsyslog_use_files_ruleset`

I could add a code that if there is no rsyslog_use_.*_ruleset in the inventory, we enable rsyslog_use_files_ruleset, which means output to the local file (e.g., /var/log/messages).  Do you think that's most customers would like to have?  But to do so, the files output needs to be configured (i.e., 30-output.files has to be in /etc/rsyslog.d).  Thinking of it, it might be safer to deploy the files output role regardless of this files configuration is given or not...
```
logging_outputs:
  - name: files-output
    type: files
```

Reply:
> > Is whether or not to use a `ruleset` an implementation detail? That is, what difference does it make to the user if the input is sent to the output using a `ruleset` or some other mechanism?
> 
> Ahhh, there's no **other** mechanism for the output... Only via rulesets.

There is no requirement to use rulesets.  If you do not specify a ruleset, the input goes into the default/general record flow (just like what happens when you use `imjournal`).  You would then have to use `if` statements to control the flow if you wanted a specific set of inputs to go to a specific set of outputs.

> 
> > That is, I would think the user would just want to specify "collect logs from these inputs, and send them to these outputs" - and that can be done without the `_ruleset` suffix e.g. `rsyslog_use_files` instead of `rsyslog_use_files_ruleset`
> 
> I could add a code that if there is no rsyslog_use_.*_ruleset in the inventory, we enable rsyslog_use_files_ruleset, which means output to the local file (e.g., /var/log/messages). Do you think that's most customers would like to have? But to do so, the files output needs to be configured (i.e., 30-output.files has to be in /etc/rsyslog.d). Thinking of it, it might be safer to deploy the files output role regardless of this files configuration is given or not...
> 
> ```
> logging_outputs:
>   - name: files-output
>     type: files
> ```

I'm not sure, but does it make a difference if you do not use rulesets at all?